Fixed #28282 ---  Fixed class-based model indexes to always name indexes on non-abstract models.

jdufresne opened the pull request:
<https://code.djangoproject.com/ticket/28282>

I also considered:

```python
if not abstract:
```

As well as removing the `if` entirely. Why not always name indexes, even on abstract models? Looking for feedback on the intended design and approach here.

timgraham commented:
I don't recall exactly (3d19d142), but there's a test failure in migrations if the condition is minimized to `if not abstract`. Did you notice it? In any case, adding a comment explaining this condition would be useful.

jdufresne commented:
I've updated the PR by removing the condition entirely. I don't believe it is necessary as the line `new_class._meta.indexes = [copy.deepcopy(idx) for idx in new_class._meta.indexes]` should be sufficient to fix [bug \#27915](https://code.djangoproject.com/ticket/27915). As each `Index` is copied, it will avoid duplicate names on extended model classes. Setting the name will always apply to the copy.

> there's a test failure in migrations if the condition is minimized to if not abstract

Jenkins is not reporting an error. Does triggering this error require a different test environment? If so please share how I can replicate.

> Adding a comment explaining this condition would be useful

I added a comment to explain the `deepcopy()`.
